doc/event-alarm-framework/events-producer.md

```diff
@@ -0,0 +1,212 @@
+# Events reported from SONiC as structured
+
+## Goals
+1. Provide a unified way for listing identified alertable events in SONiC switch.
+2. Provide a unified way for event detectors to report the events.
+3. Enforce a structured format for event data with pre-defined schema.
+4. Have the ability to stream at the max of 10K events per second to external clients.
+5. Ensure events are immutable across SONiC releases, but can be deprecated.
+6. Meet the reliability goal of 99.5% - From event generated to end client.
+7. Rate of event reporting can be in par with rate of syslog reporting.
+
+
+## Problems to solve
+The external tools that monitor system health often use syslog messages to look for events that need alert raised.
+The syslog messages being text string could potentially **change** across releases. Some log messages could get split into multiple different ones.
+This poses a challenge to the external tools as they are forced to adapt for multiple different versions and parse a message based on the OS version of the producer.
+This would become even more challenging, when we upgrade individual container images via kubernetes/app-package-manager.
+
+The tools that use syslog messages face higher latency, as they have to wait for syslog messages to arrive to an external repository. 
+This latency could run in the order of minutes.
+
+
+![image](https://user-images.githubusercontent.com/47282725/156947460-66d08b3d-c981-4413-b0d5-232643dfba01.png)
+
+
+![image](https://user-images.githubusercontent.com/47282725/159061158-30ff3c8a-5fc0-4af2-8822-6bfc67b2329c.png)
```

Comment:
>image [](http://example.com/codeflow?start=2&length=5)

In the example, I think timestamp is missing in each event json string.

Reply:
Current telemetry code that does streaming from redis,  puts its own timestamp before sending out. As receiver is working at a high speed, this would suit and as well cut down the payload size from event reporter to receiver.

But if your telemetry client design warrants, the event reporter could send the timestamp.

Reply:
Is the current scope of the project to only have one sender being the plugin binary in the host and 2 listeners being the telemetry container and the redis db container?

Reply:
The timestamp ground truth is from syslog message, which is very important. Telemetry timestamp is different.